warehouse/query-core/src/main/java/datawave/core/iterators/BoundedRangeExpansionIterator.java

```diff
@@ -0,0 +1,171 @@
+ */
+public class BoundedRangeExpansionIterator extends SeekingFilter implements OptionDescriber {
+
+    private static final Logger log = LoggerFactory.getLogger(BoundedRangeExpansionIterator.class);
+
+    public static final String START_DATE = "start.date";
+    public static final String END_DATE = "end.date";
+    public static final String DATATYPES_OPT = "dts";
+
+    private TreeSet<String> datatypes;
+    private String startDate;
+    private String endDate;
+
+    private Text prevRow;
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        if (!validateOptions(options)) {
+            throw new IllegalArgumentException("BoundedRangeExpansionIterator not configured with correct options");
+        }
+
+        String opt = options.get(DATATYPES_OPT);
+        if (StringUtils.isBlank(opt)) {
+            datatypes = new TreeSet<>();
+        } else {
+            datatypes = new TreeSet<>(Splitter.on(',').splitToList(opt));
+        }
+
+        startDate = options.get(START_DATE);
+        endDate = options.get(END_DATE) + Constants.MAX_UNICODE_STRING;
+
+        super.init(source, options, env);
+    }
+
+    @Override
+    public IteratorOptions describeOptions() {
+        IteratorOptions opts = new IteratorOptions(getClass().getName(), "Expands bounded ranges using the global index", null, null);
+        opts.addNamedOption(START_DATE, "The start date");
+        opts.addNamedOption(END_DATE, "The end date");
+        opts.addNamedOption(DATATYPES_OPT, "The set of datatypes used to filter keys");
+        return opts;
+    }
+
+    @Override
+    public boolean validateOptions(Map<String,String> options) {
+        return options.containsKey(START_DATE) && options.containsKey(END_DATE);
+    }
+
+    @Override
+    public FilterResult filter(Key k, Value v) {
+        log.trace("filter key: {}", k.toStringNoTime());
+
+        // shard + null + datatype
+        String cq = k.getColumnQualifier().toString();
+        int index = cq.indexOf('\u0000');
+        String date = cq.substring(0, index);
+
+        if (date.compareTo(startDate) < 0) {
+            log.trace("{} is before the start date {}, advancing to start date", date, startDate);
+            return new FilterResult(false, AdvanceResult.USE_HINT);
+        }
+
+        if (date.compareTo(endDate) > 0) {
+            log.trace("{} is past the end date {}, advancing to next row", date, endDate);
+            return new FilterResult(false, AdvanceResult.NEXT_ROW);
+        }
+
+        String datatype = cq.substring(index + 1);
+        if (!datatypes.isEmpty() && !datatypes.contains(datatype)) {
+            log.trace("datatype {} was filtered out, advancing to next key", datatype);
+            return new FilterResult(false, AdvanceResult.NEXT);
+        }
+
+        if (prevRow != null && prevRow.equals(k.getRow())) {
+            // should never get here
```

Comment:
I expect we would get here.  Is this now how we advance to the next value?

Reply:
Comment updated in the next commit. The iterator should only return a single key per unique row in the index, this is more of a safety check than anything else.